packages/core/streams/index.test.ts

```diff
@@ -102,6 +102,121 @@ describe('AIStream', () => {
+              headers: {
+                'x-mock-service': 'openai',
+                'x-mock-type': 'chat',
+              },
+            }),
+            {
+              onFinal() {
+                data.close();
+              },
+              experimental_streamData: true,
+            },
+          );
+
+          const response = new StreamingTextResponse(stream, {}, data);
+
+          const client = createClient(response);
+          const chunks = await client.readAll();
+
+          expect(chunks).toEqual([
+            '0:"Hello"\n',
+            '0:","\n',
+            '0:" world"\n',
+            '0:"."\n',
+          ]);
+        });
+
+        it('should send function call result', async () => {
+          const data = new experimental_StreamData();
+
+          const stream = OpenAIStream(
+            await fetch(server.api + '/mock-func-call', {
+              headers: {
+                'x-mock-service': 'openai',
+                'x-mock-type': 'func_call',
+              },
+            }),
+            {
+              onFinal() {
+                data.close();
+              },
+              experimental_streamData: true,
+            },
+          );
+
+          const response = new StreamingTextResponse(stream, {}, data);
+
+          const client = createClient(response);
+          const chunks = await client.readAll();
+
+          expect(chunks).toEqual([
+            '0:"{\\"function_call\\": {\\"name\\": \\"get_current_weather\\", \\"arguments\\": \\""\n',
+            '0:"{\\\\n"\n',
+            '0:"\\\\\\""\n',
+            '0:"location"\n',
+            '0:"\\\\\\":"\n',
+            '0:" \\\\\\""\n',
+            '0:"Char"\n',
+            '0:"l"\n',
+            '0:"ottesville"\n',
+            '0:","\n',
+            '0:" Virginia"\n',
+            '0:"\\\\\\",\\\\n"\n',
+            '0:"\\\\\\""\n',
+            '0:"format"\n',
+            '0:"\\\\\\":"\n',
+            '0:" \\\\\\""\n',
+            '0:"c"\n',
+            '0:"elsius"\n',
+            '0:"\\\\\\"\\\\n"\n',
+            '0:"}"\n',
+            '0:"\\"}}"\n',
+          ]);
+        });
+
+        it('should send text and data', async () => {
+          const data = new experimental_StreamData();
+
+          data.append({ t1: 'v1' });
+
+          const stream = OpenAIStream(
+            await fetch(server.api, {
+              headers: {
+                'x-mock-service': 'openai',
+                'x-mock-type': 'chat',
+              },
+            }),
+            {
+              onFinal() {
+                data.close();
+              },
+              experimental_streamData: true,
+            },
+          );
+
+          const response = new StreamingTextResponse(stream, {}, data);
+
+          const client = createClient(response);
+          const chunks = await client.readAll();
+
+          expect(chunks).toEqual([
+            '0:"Hello"\n',
+            '2:"[{\\"t1\\":\\"v1\\"}]"\n',
```

Comment:
I was a bit surprised that the data message shows up on the 2nd position (instead of the 1st).

Reply:
Yeah, I don't think this is expected. Great catch!

Reply:
This might be the cause: https://github.com/vercel/ai/blob/main/packages/core/streams/stream-data.ts#L33

It seems that data is only appended when the Transformer receives a regular chunk from the wrapped stream, and appends the data after. This could also potentially lead to other issues (e.g. custom data is not send immediately, but only when chunks from the underlying stream are transformed).

Reply:
After some review: this isn't a problem, as long as the order of messages with the same prefixes is kept. It can be "fixed" by moving the resulting line below the if statement